File: [40] Combination Sum II/40.combination-sum-ii.cpp
/*
 * @lc app=leetcode id=40 lang=cpp
 *
 * [40] Combination Sum II
 */

// @lc code=start
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<int> current;
        vector<vector<int>> res;
        sort(candidates.begin(), candidates.end());
        backTracking(candidates, res, current, target, 0);
        return res;
    }

    void backTracking(vector<int>& candidates, vector<vector<int>>& res, vector<int>& temp_list, int target, int start) {
        if (target == 0) {
            res.emplace_back(temp_list);
            return;
        } else if (target < 0 || start == candidates.size()) return;
        else {
            temp_list.emplace_back(candidates[start]);
            backTracking(candidates, res, temp_list, target-candidates[start], start+1);
            temp_list.pop_back();
            // after that, all combinations start with position start has been processed
            while (start < candidates.size()-1 && candidates[start] == candidates[start+1]) {
                start++;
            }
            backTracking(candidates, res, temp_list, target, start+1);
        }
    }
};
// @lc code=end

File: [40] Combination Sum II/40.combination-sum-ii.java
import java.util.ArrayList;

/*
 * @lc app=leetcode id=40 lang=java
 *
 * [40] Combination Sum II
 */

// @lc code=start
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(candidates);
        backTracking(candidates, res, new ArrayList<Integer>(), target, 0);
        return res;
    }

    public void backTracking(int[] candidates, List<List<Integer>> res, List<Integer> temp_list, int target, int start) {
        if (target == 0) {
            res.add(new ArrayList<>(temp_list));
            return;
        } else if (target < 0) {
            return;
        } else {
            for (int i = start; i < candidates.length; i++) {
                // skip repeated candidates
                // i != start means that i > start
                // position start has been processed and we have found all the combinations
                // starting from position start.
                // Therefore, if cand[start+1] == cand[start]
                // there is no need to process position start+1
                // since it will be the subset of combination starting with start position
                if (i != start && candidates[i] == candidates[i-1]) continue;
                temp_list.add(candidates[i]);
                backTracking(candidates, res, temp_list, target-candidates[i], i+1);
                temp_list.remove(temp_list.size()-1);
            }
        }
    }
}
// @lc code=end

